File: 操作系统/Windows API/Windows 程序设计/第6版/CPlusPlus/Chapter12/VisitedPageSave/VisitedPageSave/SaveStatePage.cpp
#include "pch.h"
#include "SaveStatePage.h"

using namespace VisitedPageSave;

using namespace Platform;
using namespace Platform::Collections;
using namespace Windows::Foundation::Collections;
using namespace Windows::UI::Xaml::Navigation;

Map<int, IMap<String^, Object^>^>^ SaveStatePage::pages = ref new Map<int, IMap<String^, Object^>^>();

void SaveStatePage::OnNavigatedTo(NavigationEventArgs^ args)
{
    if (args->NavigationMode == NavigationMode::New)
    {
        // Construct a dictionary key
        int pageKey = this->Frame->BackStackDepth;

        // Remove page key and higher page keys
        for (int key = pageKey; SaveStatePage::Pages->HasKey(key); key++)
            SaveStatePage::Pages->Remove(key);

        // Create a new page state dictionary and save it
        this->PageState = ref new Map<String^, Object^>();
        SaveStatePage::Pages->Insert(pageKey, this->PageState);
    }
}

File: 操作系统/Windows API/Windows 程序设计/第6版/CSharp/Chapter12/VisitedPageSave/VisitedPageSave/SaveStatePage.cs
using System.Collections.Generic;
using Windows.UI.Xaml.Controls;
using Windows.UI.Xaml.Navigation;

namespace VisitedPageSave
{
    public class SaveStatePage : Page
    {
        protected Dictionary<string, object> pageState;

        static protected Dictionary<int, Dictionary<string, object>> pages =
                                new Dictionary<int, Dictionary<string, object>>();

        protected override void OnNavigatedTo(NavigationEventArgs args)
        {
            if (args.NavigationMode == NavigationMode.New)
            {
                // Construct a dictionary key
                int pageKey = this.Frame.BackStackDepth;

                // Remove page key and higher page keys
                for (int key = pageKey; pages.Remove(key); key++) ;

                // Create a new page state dictionary and save it
                pageState = new Dictionary<string, object>();
                pages.Add(pageKey, pageState);
            }

            base.OnNavigatedTo(args);
        }
    }
}
